### DISCUSSION TITLE
Struggling to get the commadaptervda5050 to send back consecutive order topic MQTT JSON messages (same orderId)

### DISCUSSION THREAD

**Discussion by antichecker** [openTCS/opentcs#166] [Q&A]
I am working on a vda5050 based adapter which handles the communication between openTCS and the discrete event simulation software WITNESS. So far I successfully sent transport orders to the kernel via its webAPI and was able to send presumably correct `/connection` and `/state` JSON messages to the commadapter via a python script, simulating a vehicle sending these vda5050 messages.
At this point though I am struggling to make the commadapter send consecutive order messages, starting at `orderUpdateId:1`. The initial order JSON message from the commadapter is sent successfully, example:
```
{
  "headerId": 0,
  "timestamp": "2024-09-19T06:34:35.903361400Z",
  "version": "2.0.0",
  "manufacturer": "DPL",
  "serialNumber": "agv1",
  "orderId": "witness_order[1]:2024-09-19T08:22:40.239-0",
  "orderUpdateId": 0,
  "nodes": [
   {
      "nodeId": "Node01",
      "sequenceId": 0,
      "released": true,
      "actions": [],
      "nodePosition": {
        "x": 0,
        "y": 0,
        "mapId": "witness_layout1",
        "allowedDeviationXY": 0.01,
        "allowedDeviationTheta": 3.141592653589793
      }
    },
    {
      "nodeId": "Node02",
      "sequenceId": 2,
      "released": true,
      "actions": [
        {
          "actionType": "pickup",
          "actionId": "Order_destination_action_0",
          "blockingType": "NONE"
        }
      ],
      "nodePosition": {
        "x": 0,
        "y": 3,
        "mapId": "witness_layout1"
      }
    }
  ],
  "edges": [
    {
      "edgeId": "edge01",
      "sequenceId": 1,
      "released": true,
      "startNodeId": "Node01",
      "endNodeId": "Node02",
      "actions": [],
      "maxSpeed": 1
    }
  ]
}
```
As per the commadaptervda5050 runtime behaviour, the next /state message I sent mirrored the orders `orderId` and `orderUpdateId` to make the commadapter send the next order message, with `orderUpdateId:1`. 
I also filled in the `nodeStates`, `edgeStates` and `actionStates` arrays according to the received order JSON message (minimum required content):
```
{
  "headerId": 1,
  "timestamp": "2024-09-19T08:42:11.893007Z",
  "version": "2.0.0",
  "manufacturer": "DPL",
  "serialNumber": "agv1",
  "orderId": "witness_order[1]:2024-09-19T08:22:40.239-0",
  "orderUpdateId": 0,
  "lastNodeId": "Node01",
  "lastNodeSequenceId": 0,
  "nodeStates": [
    {
      "nodeId": "Node01",
      "sequenceId": 0,
      "released": true
    },
    {
      "nodeId": "Node02",
      "sequenceId": 2,
      "released": true
    }
  ],
  "edgeStates": [
    {
      "edgeId": "edge01",
      "sequenceId": 1,
      "released": true
    }
  ],
  "driving": false,
  "paused": false,
  "actionStates": [
    {
      "actionId": "Order_destination_action_0",
      "actionStatus": "WAITING",
      "actionType": "pickup"
    }
  ],
  "batteryState": {
    "batteryCharge": 100,
    "charging": false
  },
  "operatingMode": "AUTOMATIC",
  "errors": [],
  "safetyState": {
    "eStop": "NONE",
    "fieldViolation": false
  },
  "agvPosition": {
    "positionInitialized": true,
    "x": 0,
    "y": 0,
    "theta": 1.57,
    "mapId": "witness_layout1"
  }
}
```
I'm guessing there is something wrong with this `/state` message to not make the commadapter send the next 'slice' of the order. Can anyone see what's wrong?
Also the base/horizon is only one 'step' long even though the full transport order has one more node 'Node03' to drive to after 'Node02'. I configured the `vda5050:maxStepsBase` property as `3`. Any Ideas?

**Comment by swltr** [comment]
@antichecker  Which version of the VDA5050 communication adapter are you using?

In your state message, `orderId` and `orderUpdateId` look correct to me and should be sufficient for the order message to be considered accepted by the vehicle, and since `operatingMode` is `AUTOMATIC`, the driver should be able to send the next order message. Please check the kernel log for any exceptions. Maybe there is an issue here that I do not notice by merely eyeballing it.

That there is only a single step in the initial order message is correct/OK. The driver will send one order message for every step, one after another.

**Reply by antichecker** [reply]
I was using the commAdapter.vda5050 Version 2.0.
I'm not sure anymore what my problem here was but I figured out how to successfully send a /state JSON to receive the next /order JSON from openTCS. If an /order JSON includes an action at a location, the next /state message has to mirror the 'actions' properties with the property "actionStatus": "FINISHED" to tell openTCS the execution was successful and to make the commAdapter send the next order JSON message (if the transport order isn't concluded at that point).
